### ISSUE TITLE
WebSocket cannot establish secure connection

### ISSUE THREAD

**Issue by id-ilych** [cocos-creator/cocos2d-x-lite#1722]
Cannot establish `wss://` connection in mobile build.

I've created a simple project ([here](https://github.com/id-ilych/cc-web-socket-issue)) that tests both unsecure (`ws://`) and secure (`wss://`). Both work in web version but secure connection is failed in Android build. I also tried to include CA certificate and specifiy it during WebSocket creation. Doesn’t work. With extra logs I saw this socket callback invocation:

`LWS_CALLBACK_CLIENT_CONNECTION_ERROR: lws_ssl_client_connect2 failed`

Tested against `echo.websocket.org`. CA certificate is of `Let's Encrypt Authority X3`. Particular file used may be found in repo. Full log from WebSocket-libwebsockets.cpp can be found [here](https://gist.github.com/id-ilych/23032bc72cef4724906ae7712105e0b5).

**Comment by UltraNemesis** [comment]
Facing the same problem myself. Secure websocket works when I try to run the project in web browser, but no luck on the simulator. I am using TypeScript for the project btw.

I have to chose between Godot and Cocos for building some native android/iOS games and this issue is a point in favor of Godot.

**Comment by minggo** [comment]
@id-ilych @UltraNemesis 
What's your engine version?

@PatriceJiang i think you fixed the issue some days ago, right?

**Comment by id-ilych** [comment]
@minggo 
2.0.9 (https://github.com/cocos-creator/cocos2d-x-lite/releases/tag/2.0.9)

**Comment by PatriceJiang** [comment]
@id-ilych Have you try to use the cacert.pem  from project example collections?

**Comment by id-ilych** [comment]
@PatriceJiang Hi. No, I used CA of `Let's Encrypt Authority X3`. 
Where can I find project example collections?

**Comment by PatriceJiang** [comment]
You can download here https://curl.haxx.se/docs/caextract.html @id-ilych

**Comment by minggo** [comment]
@id-ilych @UltraNemesis
We will replace current implementation with okhttp soon, and will not require CA any more.

**Comment by UltraNemesis** [comment]
I am using Cocos Creator 2.1.1. The engine version is 3.9 I think. need to check.

**Comment by UltraNemesis** [comment]
How do I pass the ca cert in the typescript version? In JavaScript, 3 arguments are passed to WebSocket, In TypeScript, its only allowing 2.

**Comment by id-ilych** [comment]
@UltraNemesis I created js file to export newWebSocket method with 3 parameters to TypeScript

**Comment by UltraNemesis** [comment]
Came up with roughly a the same approach myself in between. Made a JavaScript file and exported a function from there. Picked up the ca certs from https://curl.haxx.se/docs/caextract.html

But its still not working me. 

Where did you put the cacerts.pem in the project?

**Comment by id-ilych** [comment]
@UltraNemesis Look at my sample project in the original post. I've put it in `build-templates` directory. But also modified `build.gradle` file to include it in APK.

**Comment by id-ilych** [comment]
@PatriceJiang I've updated my sample project using certificates you pointed to and not it seems to work. Cool! Thank you. Will check in my real app later and post results here.

**Comment by id-ilych** [comment]
@PatriceJiang Unfortunately on my real app and server it still doesn't work even though it also uses certificate issued by Let's Encrypt and works just fine with browser app.

Are there any estimations on migration to `okhttp`?

**Comment by minggo** [comment]
There is already a PR: https://github.com/cocos-creator/cocos2d-x-lite/pull/1731. We will merge it after testing.

**Comment by PatriceJiang** [comment]
To demonstrate how to use websocket in Creator, we need to create a builtin example project

1. open the dashboard and create a new example project
<img width="1026" alt="Screen Shot 2019-05-16 at 9 32 23 AM" src="https://user-images.githubusercontent.com/40414978/57820326-defb1800-77be-11e9-90da-a84964bddf50.png">

2. open scene "network"
<img width="748" alt="Screen Shot 2019-05-16 at 9 43 02 AM" src="https://user-images.githubusercontent.com/40414978/57820359-0b169900-77bf-11e9-9e1a-1df6683cb363.png">

3. open script `NetworkCtrl.js`
<img width="466" alt="Screen Shot 2019-05-16 at 9 45 31 AM" src="https://user-images.githubusercontent.com/40414978/57820448-6ea0c680-77bf-11e9-8176-a4a1eb3f27d0.png">

4. jump to 
<img width="932" alt="Screen Shot 2019-05-16 at 9 31 36 AM" src="https://user-images.githubusercontent.com/40414978/57820472-82e4c380-77bf-11e9-8a74-d3c4088564a0.png">
the cacert.pem is set through property, line 104

5. locate cacert.pem in file system
<img width="414" alt="Screen Shot 2019-05-16 at 9 49 02 AM" src="https://user-images.githubusercontent.com/40414978/57820552-de16b600-77bf-11e9-837f-6fbcce3bf0bd.png">
infact the CA file can be put anywhere within assets folder

6. select Canvas in node tree panel
<img width="338" alt="Screen Shot 2019-05-16 at 9 49 57 AM" src="https://user-images.githubusercontent.com/40414978/57820591-06061980-77c0-11e9-8ae1-16eb5e414656.png">

just drag the CA file to the corresponding attribute

**Comment by id-ilych** [comment]
@PatriceJiang Thank you for detailed explanation it seems more accurate than the hack I used to inlude file in build.

But I have found a culprit. If "MD5 Cache" option is set then real file name would be different from `cc.Asset.nativeUrl` therefore in `WebSocketImpl::createVhost` file is not found and ignored.

If these hashes are turned off then I can confirm - it works for my real project as well. Thank you! 😁👍

So I guess this issue could be closed and new issue with regards to "MD5 Cache" mode should be created. Should I do it?
